### DISCUSSION TITLE
new pipeline bugs

### DISCUSSION THREAD

**Discussion by megcswan** [lzachmann/models-for-missing-data#23] [General]
Hi @lzachmann. The new pipeline/grass dataset is being a little buggy for me. I have yet to get convergence, and things often blow up with this error: Error in unique(x.pred.raw) : object 'x.pred.raw' not found  Not sure if it is a problem with the data side or the pipeline side, but I wondered if you might give it a try on your side (with a covariate--I've been trying amjXEcoSite) and see how it works for you? I've also noticed that it is taking a really long time in the initial step. A single model run takes more than an hour on my machine. I'm available tomorrow except 10-12 if it would be better to talk.

**Comment by lzachmann** [comment]
I have already patched the `x.pred.raw` error, but need to look into the long startup / run-time issue. Need to do some benchmarking to determine more.

**Comment by lzachmann** [comment]
Hey Megan, the file you sent (for the previously converging "beta hurdle" model) was helpful. I'm testing things now, but one of the biggest / most critical differences between what I was _trying_ to run with the new model API and the older code involves the following....

![Screen Shot 2021-12-06 at 4 00 31 PM](https://user-images.githubusercontent.com/7574887/144936587-443d56a8-896a-4ea7-a811-d544aaea16b9.png)

The JAGS code on the right, in green, is the newer code. You'll notice that it's allowing each site, `j`, in each stratum, `k`, to have its own (random) intercept. These random site-level intercepts arise from the underlying distribution of site intercepts in that stratum with mean `mu.G0[k]` and precision (the reciprocal of the variance) `tau.G0[k]`. It's not super surprising that we are experience convergence issues if nearly all of the sites have at least _some_ perennial grass. 

The code on the left is probably closer to what we want _when_ the response we're modeling is abundant enough that most sites simply have no true zeros. In that case (the original converging code, in red), each site simply assumes the same stratum-level probability (`phi` in our JAGS code) of obtaining a true zero. The original model is specified with the following syntax in the analysis config file.

```yml
dropouts (zeros):
    - PerennialGrassCoverClass_10m (fixed)
```

The `(fixed)`, although fairly unassuming here, is critical. It basically negates the random site-level intercept term for the hurdle (the presence / absence bit). We can talk later about whether this is the most idea place to specify such a model, but for not at least it's functional. I still need to test and make the other changes we discussed, but wanted to get this down on 'paper' before it skips my mind.

**Comment by lzachmann** [comment]
Hey @megcswan turns out it was easy to convert this post from an issue to a discussion. Let's do touch base tomorrow. I'm around until 1pm MT. 

I haven't tested thoroughly yet, but will. Trying to get on top of this unscaled `X.pred` issue in our JAGS objects first. I did fix another issue for beta-hurdle models, which was to filter out 'unreasonable' combinations of model parameterizations. For example, if `PerennialGrassCoverClass_10m` or `PerennialGrassCoverClass_10m (fixed)` is included in the `dropouts (zeros)` YAML block, we cannot have a `g0-g1` model. After all, including a variable in that block implies that we want intercept only -- meaning no trend term, no additional covariates. Even if `g0-g1` is included among the `group-level effects (zeros)`, perhaps because we want to keep `g0-g1` for other responses, it just doesn't make sense for PerennialGrassCoverClass_10m. We can chat more about this tomorrow, I'm including that rather technical update here so I can use it later on for documentation purposes. At any rate, more soon.

**Comment by megcswan** [comment]
That's great, Luke. It will be good to not have the option of running an irrational model. I'm around all morning. What time works best for you?


[cid:e91861d0-6c12-4164-a017-45a096347620]

________________________________
From: Luke Zachmann ***@***.***>
Sent: Thursday, December 9, 2021 3:27 PM
To: lzachmann/models-for-missing-data ***@***.***>
Cc: Swan, Megan ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [lzachmann/models-for-missing-data] new pipeline bugs (Discussion #23)




 This email has been received from outside of DOI - Use caution before clicking on links, opening attachments, or responding.



Hey @megcswan<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmegcswan&data=04%7C01%7Cmegan_swan%40nps.gov%7C0461b63ddee2459855f708d9bb632549%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637746856774769702%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=vnyVT%2FnvLbnzCjj3DFtfOhEPH%2BX57DtUO5%2F8zZl6Tx4%3D&reserved=0> turns out it was easy to convert this post from an issue to a discussion. Let's do touch base tomorrow. I'm around until 1pm MT.

I haven't tested thoroughly yet, but will. Trying to get on top of this unscaled X.pred issue in our JAGS objects first. I did fix another issue for beta-hurdle models, which was to filter out 'unreasonable' combinations of model parameterizations. For example, if PerennialGrassCoverClass_10m or PerennialGrassCoverClass_10m (fixed) is included in the dropouts (zeros) YAML block, we cannot have a g0-g1 model. After all, including a variable in that block implies that we want intercept only -- meaning no trend term, no additional covariates. Even if g0-g1 is included among the group-level effects (zeros), perhaps because we want to keep g0-g1 for other responses, it just doesn't make sense for PerennialGrassCoverClass_10m. We can chat more about this tomorrow, I'm including that rather technical update here so I can use it later on for documentation purposes. At any rate, more soon.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Flzachmann%2Fmodels-for-missing-data%2Fdiscussions%2F23%23discussioncomment-1782440&data=04%7C01%7Cmegan_swan%40nps.gov%7C0461b63ddee2459855f708d9bb632549%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637746856774769702%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=8A6IAGKoSHJzOlnzJpXZiGbsqipOKY6UeX8q30T0MnE%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAEGVWUQAXEDKJD5VHMYVRDLUQEULLANCNFSM5JXUMCZQ&data=04%7C01%7Cmegan_swan%40nps.gov%7C0461b63ddee2459855f708d9bb632549%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637746856774769702%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=DqeH2qMUQK1QIE1lTDJRnH2BrploBBx3tmGnAPUxYxQ%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Cmegan_swan%40nps.gov%7C0461b63ddee2459855f708d9bb632549%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637746856774769702%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=X1gsDc7dh3ksucQIi4nQWQ1F6G6hlC7ePWEt%2FQBgKaw%3D&reserved=0> or Android<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Cmegan_swan%40nps.gov%7C0461b63ddee2459855f708d9bb632549%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637746856774769702%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=7Gl0judSVdfsl4vspRVV02N5A9w9PpTa6mSSq5gsEC8%3D&reserved=0>.
